internal/host/agent_server_grpc/main_linux.go

```diff
@@ -33,13 +48,192 @@ func (s *HostService) Chown(ctx context.Context, req *proto.ChownRequest) (*prot
+	if err != nil {
+		return nil, err
+	}
+
+	return &proto.LookupResponse{
+		Uid:      user.Uid,
+		Gid:      user.Gid,
+		Username: user.Username,
+		Name:     user.Name,
+		Homedir:  user.HomeDir,
+	}, nil
+}
+
+func (s *HostService) LookupGroup(ctx context.Context, req *proto.LookupGroupRequest) (*proto.LookupGroupResponse, error) {
+	name := req.Name
+	group, err := user.LookupGroup(name)
+	if err != nil {
+		return nil, err
+	}
+
+	return &proto.LookupGroupResponse{
+		Gid:  group.Gid,
+		Name: group.Name,
+	}, nil
+}
+
+func (s *HostService) Mkdir(ctx context.Context, req *proto.MkdirRequest) (*proto.Empty, error) {
+	name := req.Name
+	mode := req.Mode
+
+	if !filepath.IsAbs(name) {
+		return nil, fmt.Errorf("path must be absolute: %s", name)
+	}
+
+	if err := syscall.Mkdir(name, mode); err != nil {
+		return nil, getGrpcError(err)
+	}
+	return nil, getGrpcError(syscall.Chmod(name, mode))
+}
+
+func (s *HostService) Lstat(ctx context.Context, req *proto.LstatRequest) (*proto.LstatResponse, error) {
+	name := req.Name
+
+	if !filepath.IsAbs(name) {
+		return nil, fmt.Errorf("path must be absolute: %s", name)
+	}
+
+	var syscallStat_t syscall.Stat_t
+
+	if err := syscall.Lstat(name, &syscallStat_t); err != nil {
+		return nil, getGrpcError(err)
+	}
+	return &proto.LstatResponse{
+		Dev:     syscallStat_t.Dev,
+		Ino:     syscallStat_t.Ino,
+		Nlink:   syscallStat_t.Nlink,
+		Mode:    syscallStat_t.Mode,
+		Uid:     syscallStat_t.Uid,
+		Gid:     syscallStat_t.Gid,
+		Rdev:    syscallStat_t.Rdev,
+		Size:    syscallStat_t.Size,
+		Blksize: syscallStat_t.Blksize,
+		Blocks:  syscallStat_t.Blocks,
+		Atim: &proto.Timespec{
+			Sec:  syscallStat_t.Atim.Sec,
+			Nsec: syscallStat_t.Atim.Nsec,
+		},
+		Mtim: &proto.Timespec{
+			Sec:  syscallStat_t.Mtim.Sec,
+			Nsec: syscallStat_t.Mtim.Nsec,
+		},
+		Ctim: &proto.Timespec{
+			Sec:  syscallStat_t.Ctim.Sec,
+			Nsec: syscallStat_t.Ctim.Nsec,
+		},
+	}, nil
+
+}
+
+func (s *HostService) ReadDir(ctx context.Context, req *proto.ReadDirRequest) (*proto.ReadDirResponse, error) {
+	name := req.Name
+
+	if !filepath.IsAbs(name) {
+		return nil, fmt.Errorf("path must be absolute: %s", name)
+	}
+
+	dirEnts, err := lib.ReadDir(ctx, name)
+	if err != nil {
+		return nil, getGrpcError(err)
+	}
+
+	protoDirEnts := make([]*proto.DirEnt, 0, len(dirEnts))
+
+	for _, dirEnt := range dirEnts {
+		protoDirEnt := &proto.DirEnt{
+			Name: dirEnt.Name,
+			Type: int32(dirEnt.Type),
+			Ino:  dirEnt.Ino,
+		}
+		protoDirEnts = append(protoDirEnts, protoDirEnt)
+	}
+
+	return &proto.ReadDirResponse{
+		Entries: protoDirEnts,
+	}, nil
+
+}
+
+func (s *HostService) ReadFile(req *proto.ReadFileRequest, stream proto.HostService_ReadFileServer) error {
+	name := req.Name
+
+	if !filepath.IsAbs(name) {
+		return fmt.Errorf("path must be absolute: %s", name)
+	}
+
+	file, err := os.Open(name)
+
+	if err != nil {
+		return getGrpcError(err)
+	}
+
+	defer file.Close()
+
+	buf := make([]byte, 1024)
```

Comment:
If I understand correctly, for each 1024k of data, it'll require a round trip to the server, right? I think this is probably to small, and will likely negatively impact performance on larger files. Eg: for 1M, it'd require 1024 round trips, if we have 50ms network latency, 100ms round trip, equals to 102s to complete a transfer, even on a network with high throughput.

I'd set this for something like 8M.

Reply:
interesting point, I am **not against** increasing it however I would like to highlight that since gRPC uses HTTP2, and it has multiplexing stream by default, it will all be happening over a single TCP connection with multiplexed streams.
In practice, it allows to send multiple byte chunks of 1024 without having to wait for previous chunks to be acknowledged, aka parallelism.
Because of this, in your example @fornellas, that would take 1.2s (take it with a gran of salt, chatGPT Math)

Each of this window has a default size of 64KB but all OS nowadays has an auto scaling TCP Stack window. if you want to go deeper in this rabbit hole I found some interesting topic about[ TCP Window Scaling](https://docs.redhat.com/en/documentation/red_hat_enterprise_linux/9/html/monitoring_and_managing_system_status_and_performance/tuning-the-network-performance_monitoring-and-managing-system-status-and-performance#tcp-window-scaling_tuning-tcp-connections-for-high-throughput) in the RedHat documentation, bit too much for this topic but well there it is 😅 

Hope it makes sense.

Reply:
> In practice, it allows to send multiple byte chunks of 1024 without having to wait for previous chunks to be acknowledged, aka parallelism.

TIL. This being the case, it should not make a huge difference here then.

IIRC @ricardo6142dh was keen on actually benchmarking the HTTP agent vs gRPC agent, should be easy to make a "copy 1G file" with different sizes, so data can drive us here :-D

Reply:
Thanks for the insights! I'll increase the buffer size to 8 MB and run benchmarks with a few different sizes (e.g., 1 MB, 8 MB, and 16 MB) to measure the impact on latency and throughput. While HTTP/2 does allow for multiplexing and parallelism, reducing the number of round trips could indeed decrease latency on networks with higher response times.